pkg/virt-operator/kubevirt.go

```diff
@@ -422,6 +456,211 @@ func (c *KubeVirtController) execute(key string) error {
+			},
+		},
+		Spec: batchv1.JobSpec{
+			Template: k8sv1.PodTemplateSpec{
+
+				Spec: k8sv1.PodSpec{
+					ServiceAccountName: "kubevirt-operator",
+					RestartPolicy:      k8sv1.RestartPolicyNever,
+					Containers: []k8sv1.Container{
+						{
+							Name:            "install-strategy-upload",
+							Image:           fmt.Sprintf("%s/%s:%s", imageRegistry, "virt-operator", imageTag),
+							ImagePullPolicy: pullPolicy,
+							Command: []string{
+								"virt-operator",
+								"--dump-install-strategy",
+							},
+							Env: []k8sv1.EnvVar{
+								{
+									Name:  util.OperatorImageEnvName,
+									Value: fmt.Sprintf("%s/%s:%s", imageRegistry, "virt-operator", imageTag),
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	return job
+}
+
+func (c *KubeVirtController) garbageCollectInstallStrategyJobs() error {
+	batch := c.clientset.BatchV1()
+	jobs := c.stores.InstallStrategyJobCache.List()
+
+	for _, obj := range jobs {
+		job, ok := obj.(*batchv1.Job)
+		if !ok {
+			continue
+		}
+		if job.Status.CompletionTime == nil {
+			continue
+		}
+
+		propagationPolicy := metav1.DeletePropagationForeground
+		err := batch.Jobs(job.Namespace).Delete(job.Name, &metav1.DeleteOptions{
+			PropagationPolicy: &propagationPolicy,
+		})
+		if err != nil {
+			return err
+		}
+		log.Log.Object(job).Infof("Garbage collected completed install strategy job")
+	}
+
+	return nil
+}
+
+func (c *KubeVirtController) getInstallStrategyFromMap(version string) (*installstrategy.InstallStrategy, bool) {
+	c.installStrategyMutex.Lock()
+	defer c.installStrategyMutex.Unlock()
+
+	strategy, ok := c.installStrategyMap[version]
+	return strategy, ok
+}
+
+func (c *KubeVirtController) deleteAllInstallStrategy() error {
+	for _, obj := range c.stores.InstallStrategyConfigMapCache.List() {
+
+		configMap, ok := obj.(*k8sv1.ConfigMap)
+		if ok {
+			err := c.clientset.CoreV1().ConfigMaps(configMap.Namespace).Delete(configMap.Name, &metav1.DeleteOptions{})
+			if err != nil {
+				return err
+			}
+		}
+	}
+	c.installStrategyMutex.Lock()
+	defer c.installStrategyMutex.Unlock()
+	// reset the local map
+	c.installStrategyMap = make(map[string]*installstrategy.InstallStrategy)
+
+	return nil
+}
+
+// Loads install strategies into memory, and generates jobs to
+// create install strategies that don't exist yet.
+func (c *KubeVirtController) loadInstallStrategy(kv *v1.KubeVirt) (*installstrategy.InstallStrategy, bool, error) {
+
+	kvkey, err := controller.KeyFunc(kv)
+	if err != nil {
+		return nil, true, err
+	}
+
+	// 1. see if we already loaded the install strategy
+	strategy, ok := c.getInstallStrategyFromMap(c.config.ImageTag)
+	if ok {
+		// we already loaded this strategy into memory
+		return strategy, false, nil
+	}
+
+	// 2. look for install strategy config map in cache.
+	strategy, err = installstrategy.LoadInstallStrategyFromCache(c.stores, kv.Namespace, c.config.ImageTag)
+	if err == nil {
+		c.installStrategyMutex.Lock()
```

Comment:
i don't mind limiting the controller to a single go routine.  i'm nervous about removing the lock though since a single goroutine is not something we normally assume in a controller. it could change some day and we'd have data races.

Reply:
Would the map implementation in `sync` help?

Reply:
realistically, it won't help performance wise. the sync map would simplify the code though. I'll follow up with this.

Reply:
looking closer at sync map, i'm not happy with the type safety (or lack there of). I think i'll leave the mutex